format.swift

```diff
@@ -0,0 +1,548 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+#if canImport(Darwin)
+import Darwin.C
+#elseif canImport(Glibc)
+import Glibc
+#endif
+
+// MARK: - Logging
+
+// It’s a little unfortunate that we need to duplicate all these functions from
+// swift-syntax-dev-utils. But executing format.swift as a script is significantly
+// faster than running it from a package using `swift run` (~1s vs ~15s), which
+// makes a big difference when running the script multiple times.
+
+func logProcessCommand(executableURL: URL?, arguments: [String]?) {
+  var message = ""
+
+  if let executableURL = executableURL {
+    message += executableURL.path
+  }
+
+  if let arguments = arguments {
+    message += " \(arguments.joined(separator: " "))"
+  }
+
+  print(message)
+}
+
+// MARK: - Paths
+
+enum Paths {
+  static var gitExec: URL {
+    get throws {
+      try lookupExecutable(for: "git")
+    }
+  }
+
+  static var swiftExec: URL {
+    get throws {
+      try lookupExecutable(for: "swift")
+    }
+  }
+
+  /// The root of the swift-syntax package
+  static var packageDir: URL {
+    URL(fileURLWithPath: #file)
+      .deletingLastPathComponent()
+  }
+
+  /// The directory in which swift-format should be built.
+  static var swiftFormatBuildDir: URL {
+    packageDir
+      .appendingPathComponent(".swift-format-build")
+  }
+
+  private static var envSearchPaths: [URL] {
+    // Compute search paths from PATH variable.
+    #if os(Windows)
+    let pathSeparator: Character = ";"
+    #else
+    let pathSeparator: Character = ":"
+    #endif
+    return (paths ?? "")
+      .split(separator: pathSeparator)
+      .map(String.init)
+      .compactMap { pathString in
+        return URL(fileURLWithPath: pathString)
+      }
+  }
+
+  private static var paths: String? {
+    #if os(Windows)
+    let pathArg = "Path"
+    #else
+    let pathArg = "PATH"
+    #endif
+
+    return ProcessInfo.processInfo.environment[pathArg]
+  }
+
+  enum ExecutableLookupError: Error, CustomStringConvertible {
+    case notFound(executableName: String)
+
+    var description: String {
+      switch self {
+      case .notFound(executableName: let executableName):
+        return "Executable \(executableName) not found in PATH"
+      }
+    }
+  }
+
+  private static func lookupExecutable(for filename: String) throws -> URL {
+    let executable = envSearchPaths.map { $0.appendingPathComponent(filename) }
+      .first(where: { $0.isExecutableFile })
+    guard let executable else {
+      throw ExecutableLookupError.notFound(executableName: filename)
+    }
+    return executable
+  }
+}
+
+extension URL {
+  var isExecutableFile: Bool {
+    return (self.isFile(path) || self.isSymlink(path)) && FileManager.default.isExecutableFile(atPath: path)
```

Comment:
Do we need the file/symlink checks here? Does `isExecutableFile` return true for directories?

Reply:
Yes, exactly. FWIW, this was copied from

https://github.com/apple/swift-tools-support-core/blob/3b13e439a341bbbfe0f710c7d1be37221745ef1a/Sources/TSCBasic/FileSystem.swift#L359-L362